core/src/main/java/com/sequenceiq/cloudbreak/service/stack/InstanceMetaDataService.java

```diff
@@ -100,13 +110,39 @@ public Stack saveInstanceAndGetUpdatedStack(Stack stack, List<CloudInstance> clo
         return stack;
     }
 
-    private Map<String, String> getSubnetAzPairs(String environmentCrn) {
+    String getAzFromDiskOrNullIfRepair(Stack stack, boolean repair) {
+        String availabilityZone = null;
+        if (repair) {
+            ResourceType resourceType = getSupportedReattachableDiskType(stack);
+            if (resourceType != null) {
+                Optional<CloudResource> reattachableDiskResource = resourceRetriever.findFirstByStatusAndTypeAndStack(CommonStatus.DETACHED,
+                        resourceType, stack.getId());
```

Comment:
Sorry, maybe I am missing something but I think it is possible to have multiple `DETACHED` disk resources for a given stack that span multiple AZs (even if they all belong to the same instance group). In this case we are always picking the first such resource and using its AZ for all the upscaled instances later on.

Moreover, what if the `DETACHED` disk resources belong to (terminated) instances of different instance groups? Those IGs could be also associated with totally different AZs. The `resourceRetriever` call here does not consider IG association, so I do not see if the choice made will be always appropriate.

Reply:
Yes, you are right. It is a potential bug but I think, we need a refactor here because I don't know about the downscaled instance. Hopefully, we repair only one instance at once.

Reply:
I would be more relaxed if we found a robust approach for this. Anyway, the fix for that (along with the related comment in `AbstractResourceConnector`) may come in a later PR as well.